File: .orkestrel/campaign/docs-parity/instruments/d7/units/workflow/d7n-workflow-converge/cells.mjs
// Compares every non-`Summary` cell of the rewritten guide against the committed baseline,
// so a cut cell in a hand-rebuilt row is caught where no gate reads it.
import { readFileSync } from 'node:fs'

/** Splits one table row into its cells, keeping an escaped pipe inside a cell. */
function splitRow(line) {
	return line
		.replace(/^\|\s?/, '')
		.replace(/\s?\|$/, '')
		.split(/(?<!\\)\|/)
		.map((cell) => cell.trim())
}

/** Reads every table row of one guide, keyed by its first cell. */
function readRows(path) {
	const lines = readFileSync(path, 'utf8').split('\n')
	const rows = new Map()
	let columns
	for (let at = 0; at < lines.length; at += 1) {
		const line = lines[at]
		if (!line.startsWith('| ')) {
			columns = undefined
			continue
		}
		const cells = splitRow(line)
		if (cells.every((cell) => /^-+$/.test(cell))) continue
		if (columns === undefined) {
			columns = cells
			continue
		}
		const key = `${columns[0]}:${cells[0]}`
		const carried = rows.get(key) ?? []
		carried.push(Object.fromEntries(cells.map((cell, index) => [columns[index] ?? index, cell])))
		rows.set(key, carried)
	}
	return rows
}

const before = readRows(process.argv[2])
const after = readRows(process.argv[3])
let compared = 0
let missing = 0
const changes = []
for (const [key, rowsBefore] of before) {
	const rowsAfter = after.get(key)
	if (rowsAfter === undefined) {
		missing += 1
		changes.push(`missing after: ${key}`)
		continue
	}
	for (const [index, row] of rowsBefore.entries()) {
		const next = rowsAfter[index]
		if (next === undefined) continue
		compared += 1
		for (const [column, value] of Object.entries(row)) {
			if (['Summary', 'Behavior', 'Role', 'Value'].includes(column)) continue
			if (next[column] !== value) changes.push(`${key} [${column}]: ${value} -> ${next[column]}`)
		}
	}
}
process.stdout.write(
	`rows compared: ${compared}, rows missing after: ${missing}\n${changes.join('\n')}\n`,
)

File: .orkestrel/campaign/docs-parity/instruments/d7/units/workspace/d7n-workspace-converge/cells.py
import re, sys, pathlib
SPLIT = re.compile(r'(?<!\\)\|')
def tables(path):
    lines = pathlib.Path(path).read_text().split('\n')
    out, cur = [], []
    for line in lines:
        if line.startswith('| '):
            cur.append(line)
        else:
            if cur: out.append(cur); cur = []
    if cur: out.append(cur)
    return out
def cells(line):
    parts = SPLIT.split(line)
    return [p.strip() for p in parts[1:-1]]
def rows(path):
    d = {}
    for tbl in tables(path):
        header = cells(tbl[0])
        for line in tbl[2:]:
            c = cells(line)
            key = c[0]
            d.setdefault(key, []).append((tuple(header), tuple(c)))
    return d
a = rows(sys.argv[1]); b = rows(sys.argv[2])
compared = 0; changed = 0; missing = 0
for key, before in a.items():
    after = b.get(key)
    if after is None or len(after) != len(before):
        print(f'ROW MISSING AFTER: {key}'); missing += 1; continue
    for (h0, r0), (h1, r1) in zip(before, after):
        # compare only cells whose header is not the compared column
        keep0 = {h: v for h, v in zip(h0, r0) if h not in ('Summary', 'Behavior', 'Purpose', 'Shape / Purpose')}
        keep1 = {h: v for h, v in zip(h1, r1) if h != 'Summary'}
        for h, v in keep0.items():
            if h in keep1:
                compared += 1
                if keep1[h] != v:
                    print(f'CELL CHANGED {key} [{h}]: {v!r} -> {keep1[h]!r}'); changed += 1
print(f'rows before: {sum(len(v) for v in a.values())}, rows after: {sum(len(v) for v in b.values())}')
print(f'non-Summary cells compared: {compared}, changed: {changed}, rows missing after: {missing}')
